editor/src/messages/tool/tool_messages/select_tool.rs

```diff
@@ -334,28 +400,81 @@ impl Fsm for SelectToolFsmState {
 					// Check the last (top most) intersection layer.
 					if let Some(intersect_layer_path) = document.document_legacy.intersects_quad_root(quad, render_data).last() {
 						if let Ok(intersect) = document.document_legacy.layer(intersect_layer_path) {
-							match intersect.data {
-								LayerDataType::Text(_) => {
-									responses.push_front(ToolMessage::ActivateTool { tool_type: ToolType::Text }.into());
-									responses.push_back(TextToolMessage::Interact.into());
+							// Shallowest manipulation
+							if tool_data.selected_type == LayerSelectionBehavior::Shallowest {
+								// Double-clicking any layer within an already selected folder should select that layer
+								// Add the first layer path not already included from the intersected to our new layer path
+								let _selected_layers: Vec<_> = document.selected_layers().collect();
+								let incoming_parent = *intersect_layer_path.first().unwrap();
+								let previous_parents: Vec<_> = (0.._selected_layers.clone().len()).map(|i| &_selected_layers.clone().get(i).unwrap()[..1]).collect();
+								let mut incoming_parent_selected = false;
+								if previous_parents.contains(&&[incoming_parent].as_slice()) {
+									incoming_parent_selected = true;
 								}
-								LayerDataType::Shape(_) => {
-									responses.push_front(ToolMessage::ActivateTool { tool_type: ToolType::Path }.into());
+								if incoming_parent_selected {
+									let temp = intersect_layer_path.clone();
+									let mut intersected_layer_perm: Vec<Vec<u64>> = vec![];
+									// Permuations of intersected layer
+									for i in 1..temp.len() + 1 {
+										intersected_layer_perm.append(&mut vec![temp[..i].to_vec()]);
+									}
+									intersected_layer_perm.reverse();
+									let mut new_layer_path: Vec<u64> = vec![];
+									// Set the base layer path to the deepest layer that is currently selected
+									for perm in intersected_layer_perm {
+										for layer in _selected_layers.clone().iter() {
+											if perm == *layer.to_vec() {
```

Comment:
If you want to compare a Vec with a slice, then you can just say `my_vec == my_slice`  because [PartialEq is implemented for a Vec and a slice](https://doc.rust-lang.org/std/vec/struct.Vec.html#impl-PartialEq%3C%26%5BU%3B%20N%5D%3E-for-Vec%3CT%2C%20A%3E).